keras/engine/training.py

```diff
@@ -2404,13 +2416,18 @@ def evaluate_generator(self, generator, steps=None,
                 enqueuer.stop()
 
         if not isinstance(outs, list):
+            assert not stateful_metric_indices
             return np.average(np.asarray(all_outs),
                               weights=batch_sizes)
         else:
             averages = []
             for i in range(len(outs)):
-                averages.append(np.average([out[i] for out in all_outs],
-                                           weights=batch_sizes))
+                # [0] is the loss, the rest are metrics
+                if i == 0 or (i - 1) not in stateful_metric_indices:
```

Comment:
Unclear what the underlying logic is... either implement in a simpler/clearer way (preferred), or comment clearly.

Reply:
Oh - now I understand why other methods use Model.metrics_names rather than Model.metrics to determine the indices. Rewritten.